### DISCUSSION TITLE
Feature Request: Block, TrialInBlock, Trial Numbering Starting From 1 Instead of 0

### DISCUSSION THREAD

**Discussion by A-Ivan** [BioMotionLab/TUX#51] [Ideas]
Issue #47 

"It would be great to have the option of setting if the Block, TrialInBlock, and Trial Numbers start from either 0 or 1. Maybe this could be in the Settings file. Starting from 1 could be less confusing then starting from 0. Internally in the background code it does not matter if it is 0 or 1, but showing these as starting from 1 in the preview table/GUI and in the output file makes things easier when understanding/processing the data."

My idea is to solve this is to create an int field, which will appear in the Settings File's inspector window, asking which number the Block, TrialInBlock, and Trial Numbers should start from, which will default to 0 (the numbering currently used).

To get this to work, I need to understand how to access the following:

1. Settings File script.
2. I assume there is some master script which sets the block, trialinblock, and trial's initial value. I can have this master script read the int field from Settings File once the scene is played/executed.
3. Where is the preview table pull its information from? I will also have to set this so that the initial block, trial, etc. values are also matching the Settings File.

**Comment by AdamBebko** [comment]
I think a more robust approach would be to have an enum in the Settings file, where you select that in output Indexes start at 1 or at 0. 

then based on the value of this enum the appropriate indexes can be adjusted.

What do you think? I can try to find all the places where this would need to be updated

**Comment by AdamBebko** [comment]
Here's one spot for the runner GUI. It's in EexperimentRunnerPanel.cs and you will need to add some if checks in two spots here. It is reading the trial's data, so you'll just have to check if ColumnName ==  whichever index you're trying to modify and adjust it accordingly when its presented.

![Screen Shot 2022-01-20 at 7 18 57 PM](https://user-images.githubusercontent.com/43276915/150442859-be6f49e1-9c5d-44de-8a07-1394e3de11be.png)

**Comment by AdamBebko** [comment]
And then for the outputs and previews, they depend on this function in DataTableExtension.cs.

You'll again need to do a check if the column name equals the index columns in question and increment them by 1 if the settings file says to.

![Screen Shot 2022-01-20 at 7 58 15 PM](https://user-images.githubusercontent.com/43276915/150446348-4b82d6a7-d418-41db-b626-fd063685f7fc.png)

**Comment by AdamBebko** [comment]
As far as I know that's what's needed.

**Comment by A-Ivan** [comment]
Hi @AdamBebko 

I got the ExperimentRunnerPanel script working, as I get the setting file information from the Runner object. The problem is getting the setting information in the DataTableExtension script as there is no Runner object.

I could modify the DataTableExtension script a bit to be able to get the settings data, but I feel like there is a cleaner way to do this. 

What if when the experiment's data table is created for the first time, that is when the Settings data is checked and the initial value (Block, TrialInBlock, Trial) is set?

The only problem I see with this idea is if in other scripts you check for a specific Block, TrialInBlock, Trial value to be zero. But if this is not the case, it should be ok.

**Comment by A-Ivan** [comment]
Hi @AdamBebko 

I was able to get this working but I wasn't able to push to the dev branch or create a new branch and create a pull request, it says that permission is denied.

**Summary of the changes:**

1. Created a **ColumnNumber** class with a an enum to hold the initial Block, TrialInBlock, Trial number. This will appear in the SettingsFile.
2. **ExperimentRunnerPannel** has been changed to add an additional number to the Block, TrialInBlock, Trial during the InitRow and UpdateRowNow function calls. The UpdateRowNow function adds 1 to the CurrentTrial's data, this allows the data to be exported correctly as a csv with the correct number (starting from 1)
3. **DataTableExtension** has been changed to accept the ColumNumber as a parameter of the AsString and GetRowString functions, as the DataTableExtension does not have direct access to the runner, design, or settings file data.
3. **ExperimentGUI** has been changed to send the column number that was set in the SettingsFile when setting up the preview table.
4. **IExperimentDesignFile** and **ExperimentDesignFile2** have been changed to get the ColumnNumber from the SettingsFile.

Another solution to decrease the amount of changes would be to change the **ExperimentDesign.cs** file to:

```
DataTable BuildFinalTable(int selectedOrderIndex) {
	
	DataTable newFinalTable = baseTrialTable.Clone();

	var trialTable = RepeatAndShuffleTrialsIfNeeded();

	newFinalTable = HasBlocks ? 
		CreateTrialTableFromBlockTable(selectedOrderIndex, trialTable) 
		: RepeatAndShuffleAllTrialsIfNeeded(trialTable, newFinalTable);

	newFinalTable = TableSetup(newFinalTable);

	finalTable = newFinalTable;
	return finalTable;
}

DataTable TableSetup(DataTable oldFinalTable)
{            
	for (int rowIndex = 0; rowIndex < oldFinalTable.Rows.Count; rowIndex++)
	{
		DataRow rowData = oldFinalTable.Rows[rowIndex];

		bool trialIsComplete = (bool)rowData[designFile.GetColumnNamesSettings.Completed];

		for (int columnIndex = 0; columnIndex < oldFinalTable.Columns.Count; columnIndex++)
		{
			DataColumn column = oldFinalTable.Columns[columnIndex];
			string rowValue = rowData[column.ColumnName].ToString();
			if (column.ColumnName == "Block" || column.ColumnName == "TrialInBlock" || column.ColumnName == "Trial")
				oldFinalTable.Rows[rowIndex][column.ColumnName] = (int)oldFinalTable.Rows[rowIndex][column.ColumnName] + (int)designFile.GetColumnNumberSettings.SelectedNumber;                                            
		}
	}

	return oldFinalTable;
}
```

I tried this but this would make the experiment start from trial 2 (originally trial 1) instead of 1 (originally trial 0), and I could not figure out where to fix it. There are also places that add 1 to the current trial and block to print out information, those would also have to be changed.

Concerning the names of classes, they probably aren't the best, I just used these as I couldn't think of a better one.

If you have any questions about anything let me know.

**Reply by AdamBebko** [reply]
I looked into the pull request problem. Everything should be ok. Did you initiate the pull request from a fork? That's what you need to do for it to work since I don't want to allow random people to make changes without approval by us first.

So it's been a while, but basically I think what you need to do is create a fork of the dev branch, then pull request from your fork to my dev branch.  @[DerMilchmann](https://github.com/DerMilchmann) did one successfully so maybe you can ask him if you run into trouble. 

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork

**Reply by AdamBebko** [reply]
Anyway, your changes look perfect, happy to review your pull request.
